File: backend/src/BloodDonationNetwork.Domain/Entities/BloodRequest.cs
using System;
using BloodDonationNetwork.Domain.Enums;

namespace BloodDonationNetwork.Domain.Entities;

public class BloodRequest
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public Guid RequesterId { get; set; }
    public Guid OrganizationId { get; set; }
    public BloodType BloodType { get; set; }
    public int UnitsRequested { get; set; }
    public RequestUrgency Urgency { get; set; }
    public BloodRequestStatus Status { get; set; } = BloodRequestStatus.Pending;
    public string HospitalName { get; set; } = string.Empty;
    public double Latitude { get; set; }
    public double Longitude { get; set; }
    public string Notes { get; set; } = string.Empty;
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime? FulfilledAt { get; set; }
    public DateTime? ClosedAt { get; set; }

    // Navigation properties
    public User? Requester { get; set; }
    public Organization? Organization { get; set; }
}

File: mobile/lib/features/blood_requests/blood_request.dart
/// Mirrors the backend RequestResponseDto.
class BloodRequest {
  const BloodRequest({
    required this.id,
    required this.requesterId,
    required this.organizationId,
    required this.bloodType,
    required this.unitsRequested,
    required this.urgency,
    required this.status,
    required this.hospitalName,
    required this.latitude,
    required this.longitude,
    required this.notes,
    required this.createdAt,
    this.fulfilledAt,
    this.closedAt,
  });

  final String id;
  final String requesterId;
  final String organizationId;
  final String bloodType;
  final int unitsRequested;
  final String urgency;
  final String status;
  final String hospitalName;
  final double latitude;
  final double longitude;
  final String notes;
  final DateTime createdAt;
  final DateTime? fulfilledAt;
  final DateTime? closedAt;

  factory BloodRequest.fromJson(Map<String, dynamic> json) => BloodRequest(
        id: json['id'] as String,
        requesterId: json['requesterId'] as String,
        organizationId: json['organizationId'] as String,
        bloodType: json['bloodType'] as String,
        unitsRequested: (json['unitsRequested'] as num).toInt(),
        urgency: json['urgency'] as String,
        status: json['status'] as String,
        hospitalName: json['hospitalName'] as String,
        latitude: (json['latitude'] as num).toDouble(),
        longitude: (json['longitude'] as num).toDouble(),
        notes: json['notes'] as String,
        createdAt: DateTime.parse(json['createdAt'] as String),
        fulfilledAt: _parseDateTime(json['fulfilledAt']),
        closedAt: _parseDateTime(json['closedAt']),
      );

  static DateTime? _parseDateTime(Object? value) => value == null
      ? null
      : DateTime.parse(value as String);
}